go/libraries/doltcore/table/typed/noms/range_reader.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package noms
+
+import (
+	"context"
+	"io"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/row"
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/schema"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+// InRangeCheck is a call made as the reader reads through values to check that the next value
+// being read is in the range]
+type InRangeCheck func(tuple types.Tuple) (bool, error)
```

Comment:
What's the advantage to having a range check as opposed to upper and lower bound?

Reply:
There is no advantage yet, but when dealing with multi column keys it provides more flexibility, and it makes it trivial to write a default check function which just compares against an end value if that's the behavior you want.